### PR TITLE
Update Web Terminal custom resources to use v1alpha2

### PR DESCRIPTION
**Pull request by amisevsk**
### Description

This PR adds support for `v1alpha2` Web Terminal custom resources (DevWorkspaces) via a feature flag. If `v1alpha2` DevWorkspaces are available on the cluster, the console frontend will create those instead of current `v1alpha1` Web Terminals to continue to support web terminal functionality. 

Starting with OpenShift 4.8, we intend to make the Web Terminal operator depend on the DevWorkspace Operator via OLM. This change requires changes to the Web Terminal custom resource that gets created. Currently (i.e. OpenShift 4.7) the Web Terminal operator is deploying an old version of the DevWorkspace Operator under the hood.

### Related PRs
- Reconfigure WTO as depending on the DevWorkspace Operator: https://github.com/redhat-developer/web-terminal-operator/pull/80
- Begin building standalone operator for DevWorkspaces: https://github.com/devfile/devworkspace-operator/pull/438

### Testing this PR
**Note**: if the cluster being used for testing is OpenShift 4.7, it's necessary to include changes from https://github.com/openshift/console-operator/pull/513, which changed the backend console RBAC. The easiest way to do this is to run 
```
watch -n 1 kubectl apply -f manifests/03-rbac-role-cluster.yaml
```
from the console-operator repo while testing.

#### Testing that current functionality still works:
1. Install the Web Terminal Operator (version 1.2.1) via the OperatorHub
    - The 1.2.1 Web Terminal Operator uses the older v1alpha1 spec.
2. Check that terminals work for both cluster-admin and regular users.

#### Testing that new functionality works as expected:
1. There's no upgrade path from WTO 1.2.1 to 1.3.0, as the switch from a complete operator to one that depends on another operator is incompatible. As a result, existing instances of WTO 1.2.1 must be fully uninstalled:

    ```bash
    # 1. Ensure that all DevWorkspace Custom Resources are removed to avoid issues with finalizers
    # make sure depending objects are clean up as well
    kubectl delete devworkspaces.workspace.devfile.io --all-namespaces --all --wait
    kubectl delete workspaceroutings.controller.devfile.io --all-namespaces --all --wait
    kubectl delete components.controller.devfile.io --all-namespaces --all --wait
    # 2. Uninstall the Operator
    kubectl delete subscriptions.operators.coreos.com web-terminal -n openshift-operators --ignore-not-found
    WTO_CSV=$(kubectl get csv -o=json | jq -r '[.items[] | select (.metadata.name | contains("web-terminal.v1"))][0].metadata.name')
    kubectl delete csv ${WTO_CSV} -n openshift-operators
    # 3. Remove CRDs
    kubectl delete customresourcedefinitions.apiextensions.k8s.io workspaceroutings.controller.devfile.io
    kubectl delete customresourcedefinitions.apiextensions.k8s.io components.controller.devfile.io
    kubectl delete customresourcedefinitions.apiextensions.k8s.io devworkspaces.workspace.devfile.io
    # 4. Remove DevWorkspace Webhook Server Deployment itself
    kubectl delete deployment/devworkspace-webhook-server -n openshift-operators
    # 5. Remove lingering service, secrets, and configmaps
    kubectl delete all --selector app.kubernetes.io/part-of=devworkspace-operator,app.kubernetes.io/name=devworkspace-webhook-server
    kubectl delete serviceaccounts devworkspace-webhook-server -n openshift-operators
    kubectl delete configmap devworkspace-controller -n openshift-operators
    kubectl delete clusterrole devworkspace-webhook-server
    kubectl delete clusterrolebinding devworkspace-webhook-server
    # 6. Remove mutating/validating webhooks configuration
    kubectl delete mutatingwebhookconfigurations controller.devfile.io
    kubectl delete validatingwebhookconfigurations controller.devfile.io
    ```
2. Apply custom CatalogSource to provide new version of Web Terminal Operator and dependency DevWorkspace Operator:

    ```yaml
    cat <<EOF | oc apply -f -
    apiVersion: operators.coreos.com/v1alpha1
    kind: CatalogSource
    metadata:
      name: custom-web-terminal-catalog
      namespace: openshift-marketplace
    spec:
      sourceType: grpc
      image: quay.io/amisevsk/devworkspace-demo-index:openshift-console
      publisher: Red Hat
      displayName: Web Terminal Operator Catalog
    EOF
    ```
3. Install Web Terminal Operator 1.3.0 from OperatorHub (it will be a separate entry, not in the default catalog)
4. Check that 1. Web Terminals work as expected for cluster-admin and regular users, and 2. that the CR created for Web Terminals has the `v1alpha2` version (`oc get dw -o yaml` from a web terminal)

### Cleanup
<details><summary>Cleaning up WTO 1.3.0 after testing</summary>
<p>

```bash
# -- 1 -- Uninstall WTO subscription and related CSV
kubectl delete subscriptions.operators.coreos.com web-terminal -n openshift-operators --ignore-not-found
WTO_CSV=$(kubectl get csv -o=json | jq -r '[.items[] | select (.metadata.name | contains("web-terminal.v1"))][0].metadata.name')
kubectl delete csv ${WTO_CSV} -n openshift-operators

# -- 2 -- Uinstall DWO subscription and related CSV
kubectl delete subscriptions.operators.coreos.com devworkspace-operator -n openshift-operators --ignore-not-found
DWO_CSV=$(kubectl get csv -o=json | jq -r '[.items[] | select (.metadata.name | contains("devworkspace-operator.v0"))][0].metadata.name')
kubectl delete csv ${DWO_CSV} -n openshift-operators

# -- 3 -- Make sure all related DWO objects are removed
  # 1. Ensure that all DevWorkspace Custom Resources are removed to avoid issues with finalizers
  # make sure depending objects are clean up as well
  kubectl delete devworkspaces.workspace.devfile.io --all-namespaces --all --wait
  kubectl delete devworkspaceroutings.controller.devfile.io --all-namespaces --all --wait
        kubectl delete devworkspacetemplates.workspace.devfile.io --all-namespaces --all --wait
  # 3. Remove CRDs
  kubectl delete customresourcedefinitions.apiextensions.k8s.io devworkspaceroutings.controller.devfile.io
  kubectl delete customresourcedefinitions.apiextensions.k8s.io devworkspacetemplates.controller.devfile.io
  kubectl delete customresourcedefinitions.apiextensions.k8s.io devworkspaces.workspace.devfile.io
  # 4. Remove DevWorkspace Webhook Server Deployment itself
  kubectl delete deployment/devworkspace-webhook-server -n openshift-operators
  # 5. Remove lingering service, secrets, and configmaps
  kubectl delete all --selector app.kubernetes.io/part-of=devworkspace-operator,app.kubernetes.io/name=devworkspace-webhook-server
  kubectl delete serviceaccounts devworkspace-webhook-server -n openshift-operators
  kubectl delete configmap devworkspace-controller -n openshift-operators
  kubectl delete clusterrole devworkspace-webhook-server
  kubectl delete clusterrolebinding devworkspace-webhook-server
  # 6. Remove mutating/validating webhooks configuration
  kubectl delete mutatingwebhookconfigurations controller.devfile.io
  kubectl delete validatingwebhookconfigurations controller.devfile.io
```

</p>
</details>

### CODE REVIEW

**Review by sahil143 — APPROVED** [review]
/lgtm

tested web terminal operator v1.3.0 on a 4.8 cluster. works fine

**Comment by sleshchenko** [comment]
Only PR author(or repo maintainer) is able to modify PR description, so providing newer info here and it'll be moved to PR description later by one who has access.

So, PR is reworked to make OpenShift Console compatible with WTO 1.2(brings DevWorkspace v1alpha1) and WTO 1.3 (brings DevWorkspace v1alpha2).

It's needed since we're not able to provide WTO 1.3 release on the same day as OpenShift 4.8 is released.
+ WTO 1.2 -> WTO 1.3 automatic upgrade is not available, so we'll have some OS 4.8 cluster with WTO 1.2 installed.

So:
- backend changes: for simplification, backend always uses v1alpha1 and lets K8s API convert v2 -> v1 when WTO 1.3 is installed;
- frontend according to the discovered DevWorkspace API Version use one or another DevWorkspace Configuration. Note that DevWorkspace CR created for WTO 1.2 should not be migrated since WTO 1.2 -> 1.3 upgrade is not supported. Administrators must remove WTO 1.2 in a proper documented way, where they remove all related CRs and CRDs, and only then install WTO 1.3.

## How to test this PR

``` bash
# -- 1 -- build and run console from PR.
# I built image and used the following commands on OpenShift 4.7
oc patch consoles.operator.openshift.io cluster --patch '{ "spec": { "managementState": "Unmanaged" } }' --type=merge
oc set image deploy console console=quay.io/sleshche/openshift-console:os4.8 -n openshift-console
oc rollout status -w deploy/console -n openshift-console
# if you apply it on OpenShift 4.8, also need to apply console operator changes:
# from console-operator folder
watch -n 1 kc apply -f manifests/03-rbac-role-cluster.yaml


# -- 2 --
# install WTO 1.2 via operatorhub and after it's done make sure terminals works 
# DevWorkspace in old format are created

# -- 3 --
# uninstall WTO 1.2 in a proper way. See Operator description on OpenShift Console or do:

    # 1. Ensure that all DevWorkspace Custom Resources are removed to avoid issues with finalizers
	# make sure depending objects are clean up as well
	kubectl delete devworkspaces.workspace.devfile.io --all-namespaces --all --wait
	kubectl delete workspaceroutings.controller.devfile.io --all-namespaces --all --wait
	kubectl delete components.controller.devfile.io --all-namespaces --all --wait
	# 2. Uninstall the Operator
	kubectl delete subscriptions.operators.coreos.com web-terminal -n openshift-operators --ignore-not-found
WTO_CSV=$(kubectl get csv -o=json | jq -r '[.items[] | select (.metadata.name | contains("web-terminal.v1"))][0].metadata.name')      
        kubectl delete csv ${WTO_CSV} -n openshift-operators
	# 3. Remove CRDs
	kubectl delete customresourcedefinitions.apiextensions.k8s.io workspaceroutings.controller.devfile.io
	kubectl delete customresourcedefinitions.apiextensions.k8s.io components.controller.devfile.io
	kubectl delete customresourcedefinitions.apiextensions.k8s.io devworkspaces.workspace.devfile.io
	# 4. Remove DevWorkspace Webhook Server Deployment itself
	kubectl delete deployment/devworkspace-webhook-server -n openshift-operators
	# 5. Remove lingering service, secrets, and configmaps
	kubectl delete all --selector app.kubernetes.io/part-of=devworkspace-operator,app.kubernetes.io/name=devworkspace-webhook-server
	kubectl delete serviceaccounts devworkspace-webhook-server -n openshift-operators
	kubectl delete configmap devworkspace-controller -n openshift-operators
	kubectl delete clusterrole devworkspace-webhook-server
	kubectl delete clusterrolebinding devworkspace-webhook-server
	# 6. Remove mutating/validating webhooks configuration
	kubectl delete mutatingwebhookconfigurations controller.devfile.io
	kubectl delete validatingwebhookconfigurations controller.devfile.io


# -- 4 --
# Apply Custom Catalog Source to get WTO 1.3 build from main branch of https://github.com/redhat-developer/web-terminal-operator
cat <<EOF | oc apply -f -
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: custom-web-terminal-catalog
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: quay.io/amisevsk/devworkspace-demo-index:openshift-console
  publisher: Red Hat
  displayName: Web Terminal Operator Catalog
EOF

# -- 5 --
# install WTO 1.3 with operatorhub and make sure terminals are still working.
# DevWorkspaces in a new format are created.
```

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx [hunk 1]
```diff
@@ -156,7 +158,7 @@ const CloudShellExec: React.FC<CloudShellExecProps> = ({
         setWsOpen(false);
 
         // Check the Cloud Shell to see if it has any hints as to why the terminal connection was closed
-        const cloudShellCR = getCloudShellCR(workspaceName, namespace);
+        const cloudShellCR = getCloudShellCR(workspaceModel, workspaceName, namespace);
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx:161]: Please add `workspaceModel` to the dependency array. It's present below the return statement. line 208

**Review by sahil143 — CHANGES_REQUESTED** [review]
frontend is failing because of linter errors.

try command `yarn run lint`

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx (hunk 1 above)
Comment (sleshchenko) [reply frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx:161]: Thanks. Fixed this as well as other linter errors.

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellTerminal.tsx [hunk 2]
```diff
@@ -40,7 +41,17 @@ const CloudShellTerminal: React.FC<CloudShellTerminalProps &
     verb: 'create',
     resource: 'pods',
   });
-  const [workspace, loaded, loadError] = useCloudShellWorkspace(user, isAdmin, namespace);
+  const [isv1Alpha2Available, isv1alpha2CheckLoading] = CheckV1alpha2CRDAvailability();
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/cloud-shell/CloudShellTerminal.tsx:44]:
```suggestion
  const [isv1Alpha2Available, isv1alpha2CheckLoading] = useV1alpha2CRDAvailability();
```

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellTerminal.tsx [hunk 3]
```diff
@@ -40,7 +41,17 @@ const CloudShellTerminal: React.FC<CloudShellTerminalProps &
...
+  let workspaceModel = WorkspaceModel;
+  if (!isv1Alpha2Available) {
+    workspaceModel = v1alpha1WorkspaceModel;
+  }
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/cloud-shell/CloudShellTerminal.tsx:48]:
```suggestion
  const workspaceModel = React.useMemo(() => (!isv1Alpha2Available ? v1alpha1WorkspaceModel : WorkspaceModel), [isv1Alpha2Available])
```

File: frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts [hunk 4]
```diff
@@ -1,31 +1,50 @@
...
+  plugin: {
+    kubernetes: {
+      name: string;
+      namespace?: string;
+    };
   };
+};
+
+type DevWorkspaceTemplateSpec = {
+  components: v1alpha1Component | v1alpha2Component[];
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts:26]:
```suggestion
  components: v1alpha1Component[] | v1alpha2Component[];
```

File: frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts [hunk 5]
```diff
@@ -36,37 +55,85 @@ export const CLOUD_SHELL_PROTECTED_NAMESPACE = 'openshift-terminal';
...
-export const newCloudShellWorkSpace = (name: string, namespace: string): CloudShellResource => ({
-  apiVersion: 'workspace.devfile.io/v1alpha1',
-  kind: 'DevWorkspace',
-  metadata: {
-    name,
-    namespace,
-    labels: {
-      [CLOUD_SHELL_LABEL]: 'true',
+const v1alpha1DevworkspaceComponent = () => {
+  return [
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts:59]: no need for method, can be a constant
```suggestion
const v1alpha1DevworkspaceComponent = [
```

File: frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts [hunk 6]
```diff
@@ -36,37 +55,85 @@ export const CLOUD_SHELL_PROTECTED_NAMESPACE = 'openshift-terminal';
...
+        id: 'redhat-developer/web-terminal/latest',
+      },
     },
-    annotations: {
-      [CLOUD_SHELL_RESTRICTED_ANNOTATION]: 'true',
+  ];
+};
+
+const devWorkspaceComponent = () => {
+  return [
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts:70]: same as above.. can be a constant

File: frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts [hunk 7]
```diff
@@ -36,37 +55,85 @@ export const CLOUD_SHELL_PROTECTED_NAMESPACE = 'openshift-terminal';
...
+      },
+      annotations: {
+        [CLOUD_SHELL_RESTRICTED_ANNOTATION]: 'true',
+      },
+    },
+    spec: {
+      started: true,
+      routingClass: 'web-terminal',
+      template: {
+        components: [],
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts:114]:
```suggestion
        components: version === 'v1alpha1 ? v1alpha1DevworkspaceComponent : devWorkspaceComponent,
```

File: frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts [hunk 8]
```diff
@@ -36,37 +55,85 @@ export const CLOUD_SHELL_PROTECTED_NAMESPACE = 'openshift-terminal';
...
+      },
     },
-  },
-});
+  };
+  if (version === 'v1alpha1') {
+    devworkspace.spec.template.components = v1alpha1DevworkspaceComponent();
+  } else {
+    devworkspace.spec.template.components = devWorkspaceComponent();
+  }
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts:122]: can be removed
```suggestion
```

File: frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts [hunk 9]
```diff
@@ -36,37 +55,85 @@ export const CLOUD_SHELL_PROTECTED_NAMESPACE = 'openshift-terminal';
...
 export const startWorkspace = (workspace: CloudShellResource) => {
-  return k8sPatch(WorkspaceModel, workspace, [
+  // Check the version so we know what workspace model to use for starting the workspace
+  const groupVersion = workspace.apiVersion.split('/');
+  const version = groupVersion[1];
+
+  let workspaceModel = WorkspaceModel;
+  if (version === 'v1alpha1') {
+    workspaceModel = v1alpha1WorkspaceModel;
+  }
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts:134]:
```suggestion
  const workspaceModel = version === 'v1alpha1' ? v1alpha1WorkspaceModel :  WorkspaceModel;
```

**Review by sahil143 — CHANGES_REQUESTED** [review]

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellTerminal.tsx (hunk 3 above)
Comment (christianvogt) [reply frontend/packages/console-app/src/components/cloud-shell/CloudShellTerminal.tsx:48]: A memo here is overkill as there's no computed value here that requires a stable reference. It's simply an if condition to reassign the `workspaceModel` value.

File: frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts [hunk 10]
```diff
@@ -96,6 +163,29 @@ export const sendActivityTick = (workspaceName: string, namespace: string): void
...
+};
+
+// Check to see if v1alpha2 devworkspace crds are available on the cluster.
+// If they are not available or if an error occurs when finding them then report that they aren't available
+export const CheckV1alpha2CRDAvailability = () => {
+  const [loading, setLoading] = useSafetyFirst(true);
+  const [isAvailable, setAvailable] = useSafetyFirst(false);
+
+  React.useEffect(() => {
+    k8sGet(WorkspaceModel)
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts:177]: Are you trying to fetch a CRD here? This won't work for a normal user.
You should be using feature flags which makes use of API discovery.
You can see how we create the following feature flag in the `console-extensions.json`:
```
  {
    "type": "console.flag/model",
    "properties": {
      "flag": "DEVWORKSPACE",
      "model": {
        "group": "workspace.devfile.io",
        "version": "v1alpha1",
        "kind": "DevWorkspace"
      }
    }
  },
```

Then you can use `useFlag` hook to check the enablement.

**Review by christianvogt — COMMENTED** [review]
I think you'll need a new feature flag for the new version of the model because if there are v1alpha2 and v1alpha1 versions then you need to enable terminal if either of these CRDs is installed.

File: frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts (hunk 10 above)
Comment (christianvogt) [reply frontend/packages/console-app/src/components/cloud-shell/cloud-shell-utils.ts:177]: Ok k8sGet will fetch all resources of that kind.
That's not efficient. So using feature flags is still a better solution.

**Review by sahil143 — APPROVED** [review]
/lgtm

@JPinkney Thanks for the changes. PR looks good overall.

Verified on a 4.8 cluster and both versions work fine.
Noticed one issue, once v1.3 was uninstalled I wasn't able to install v1.2 again but that might be because of missing something in the uninstallation steps for v1.3.
cc @sleshchenko

**Pull request merged**